apis/v1alpha1/nginxproxy_types.go

```diff
@@ -114,3 +120,69 @@ type TelemetryExporter struct {
+	Mode *RewriteClientIPModeType `json:"mode,omitempty"`
+
+	// SetIPRecursively configures whether recursive search is used when selecting the client's address from
+	// the X-Forwarded-For header. It is used in conjunction with TrustedAddresses.
+	// If enabled, NGINX will recurse on the values in X-Forwarded-Header from the end of array
+	// to start of array and select the first untrusted IP.
+	// For example, if X-Forwarded-For is [11.11.11.11, 22.22.22.22, 55.55.55.1],
+	// and TrustedAddresses is set to 55.55.55.1, NGINX will rewrite the client IP to 22.22.22.22.
+	// If disabled, NGINX will select the IP at the end of the array.
+	// In the previous example, 55.55.55.1 would be selected.
+	// Sets NGINX directive real_ip_recursive: https://nginx.org/en/docs/http/ngx_http_realip_module.html#real_ip_recursive
+	//
+	// +optional
+	SetIPRecursively *bool `json:"setIPRecursively,omitempty"`
+
+	// TrustedAddresses specifies the addresses that are trusted to send correct client IP information.
+	// If a request comes from a trusted address, NGINX will rewrite the client IP information,
+	// and forward it to the backend in the X-Forwarded-For* and X-Real-IP headers.
+	// If the request does not come from a trusted address, NGINX will not rewrite the client IP information.
+	// Addresses must be provided as CIDR blocks or IP addresses: 10.0.0.0, 192.33.21/24, fe80::1/128.
+	// To trust all addresses (not recommended for production), set to 0.0.0.0/0.
+	// If no addresses are provided, NGINX will not rewrite the client IP information.
+	// Sets NGINX directive set_real_ip_from: https://nginx.org/en/docs/http/ngx_http_realip_module.html#set_real_ip_from
+	// This field is required if mode is set.
+	// +kubebuilder:validation:MaxItems=16
+	// +listType=set
+	//
+	//
+	// +optional
+	TrustedAddresses []TrustedAddress `json:"trustedAddresses,omitempty"`
+}
+
+// RewriteClientIPModeType defines how NGINX Gateway Fabric will determine the client's original IP address.
+// +kubebuilder:validation:Enum=ProxyProtocol;XForwardedFor
+type RewriteClientIPModeType string
+
+const (
+	// RewriteClientIPModeProxyProtocol configures NGINX to accept PROXY protocol and
+	// set the client's IP address to the IP address in the PROXY protocol header.
+	// Sets the proxy_protocol parameter on the listen directive of all servers and sets real_ip_header
+	// to proxy_protocol: https://nginx.org/en/docs/http/ngx_http_realip_module.html#real_ip_header.
+	RewriteClientIPModeProxyProtocol RewriteClientIPModeType = "ProxyProtocol"
+
+	// RewriteClientIPModeXForwardedFor configures NGINX to set the client's IP address to the
+	// IP address in the X-Forwarded-For HTTP header.
+	// https://nginx.org/en/docs/http/ngx_http_realip_module.html#real_ip_header.
+	RewriteClientIPModeXForwardedFor RewriteClientIPModeType = "XForwardedFor"
+)
+
+// TrustedAddress is a string value representing a CIDR block or an IP address.
+// Examples: 10.0.0.2/32, 10.0.0.1, fe80::1/128, ::1/24.
```

Comment:
fe80::1/128 does not match the regex.

Reply:
@kate-osborn The regex to match ipv4/ipv6 cidr OR address isn't very clear (or maybe even possible). What I was thinking is we match ipv4/ipv6 address with optional trailing `/` and up to 3 digits. This would obviously allow ipv4 with a 3 digit cidr, but our Go validation would catch that.

Any thoughts on if there might be a better solution? I considered if we split address and CIDR into two separate fields, but I wonder if that just complicates the API too much versus just a basic string array.

Reply:
OpenAPI supports ipv4 and ipv6 formatting for strings, and maybe in the future will support our needs: https://github.com/OAI/OpenAPI-Specification/issues/3692

Reply:
I think if we were to choose one, it should be CIDR for the flexibility. But I don't think there would be a very clean regex for that.

Reply:
The gateway API had to add some custom logic to generate that json schema linked above: https://github.com/kubernetes-sigs/gateway-api/blob/cb846ce842f47074781de7ebfbb1080f3c471f2f/pkg/generator/main.go#L142

Reply:
We could follow the pattern established in the gateway API:

```
type Address struct {
    Type AddressType
    Value string
}

type AddressType string

const (
    IPAddressType AddressType = IPAddress
)

```

and start by supporting IP addresses, which are easier to validate with regex. We allow 16 addresses, which should be enough for most use cases. If we get feedback asking for CIDR support, we can easily add that using the above pattern and apply a separate regex. 

I suppose we could start with CIDR instead and leave the bulk of the validation up to the go code...I'm just worried about the user experience.

Reply:
Even the regex examples I've seen for just checking ipv4 or ipv6 are fairly complex, though maybe we could do the same workaround that the Gateway API folks did by adding `format: ipv4` and `format: ipv6` (assuming that this does validation and isn't just documentation).

I was hesitant about moving away from a string to a strong type, just because a string is a better UX, but this could be better for parsing I guess.

If we had a CIDR type, we could also add some basic regex to ensure it meets some criteria.

I still feel like we need to support CIDR. IP addresses change, and having to update this field every time your LB IP address changes is a pain, whereas the network CIDR shouldn't change.

Reply:
> (assuming that this does validation and isn't just documentation).

It does validate, I checked 🙂 

> I still feel like we need to support CIDR. IP addresses change, and having to update this field every time your LB IP address changes is a pain, whereas the network CIDR shouldn't change.

Fair point. I'm fine with starting with CIDR and using a basic regex as long as that regex doesn't prevent valid CIDRs.

> I was hesitant about moving away from a string to a strong type, just because a string is an easier UX, but this could be better for parsing I guess.

It is a more complex UX, but it allows us to add other address types easily in the future. For example, we could add hostname support since `set_real_ip_from` works with hostnames, too.

Reply:
> set_real_ip_from works with hostnames, too

Oh I didn't know that. I think that solidifies using the struct instead of a string then.

@salonichf5 So I think let's make an update to this to support the struct proposed by Kate above, but for now let's just support CIDR. We can create a story for supporting IP addresses. I would say we could support IP addresses now, but it'll be a little work to generate the OpenAPI schema for the ipv4 and ipv6 types like the Gateway API does, so maybe we just do that another time to get this PR through.

Reply:
I've incorporated these changes.

Reply:
all conformance and functional tests were failing saying a struct cannot have type `set`, should be atomic